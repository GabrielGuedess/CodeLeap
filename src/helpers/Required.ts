export type Required<T> = {
  [P in keyof T]-?: T[P];
};
