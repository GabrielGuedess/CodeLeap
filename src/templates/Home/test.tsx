import { screen, waitFor } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';

import { renderWithTheme } from 'utils/tests/helpers';

import { Home } from '.';

const createPostMock = jest.fn();

let isLoading = false;
let posts = [
  {
    id: '1',
    title: 'Title',
    content: 'Content',
    username: 'GabrielGuedess',
    createdAt: '2023-04-07T00:00:00.000Z',
  },
];

jest.mock('@formkit/auto-animate/react', () => ({
  useAutoAnimate: () => [null],
}));

jest.mock('hooks/usePost', () => ({
  usePost: () => ({
    posts,
    createPost: createPostMock,
    isLoading,
  }),
}));

jest.mock('hooks/useAuth', () => ({
  useAuth: () => ({
    username: 'GabrielGuedess',
  }),
}));

jest.mock('components/molecules/Header', () => ({
  __esModule: true,
  Header: function Header({ children }: { children: React.ReactNode }) {
    return <div data-testid="Header Mock">{children}</div>;
  },
}));

jest.mock('components/organisms/Post', () => ({
  __esModule: true,
  Post: function Post({ children }: { children: React.ReactNode }) {
    return <div data-testid="Post Mock">{children}</div>;
  },
}));

describe('<Home />', () => {
  it('should render the home correctly', () => {
    // Arrange
    const { container } = renderWithTheme(<Home />);

    // Assert
    expect(screen.getByTestId('Header Mock')).toBeInTheDocument();
    expect(screen.getByTestId('Post Mock')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the empty component', async () => {
    posts = [];

    renderWithTheme(<Home />);

    // Assert
    expect(
      screen.getByRole('heading', { name: 'Network is Empty' }),
    ).toBeInTheDocument();
  });

  it('should render the loading component', async () => {
    isLoading = true;

    renderWithTheme(<Home />);

    // Assert
    expect(screen.getByLabelText('Spinner')).toBeInTheDocument();
  });

  it('should create new post when click on create button', async () => {
    renderWithTheme(<Home />);

    // Act
    await userEvent.type(screen.getByLabelText('Title'), 'Gabriel');
    await userEvent.type(screen.getByLabelText('Content'), 'Text');

    await userEvent.click(screen.getByRole('button', { name: 'CREATE' }));

    // Assert
    await waitFor(() => {
      expect(createPostMock).toHaveBeenCalled();
    });
  });
});
