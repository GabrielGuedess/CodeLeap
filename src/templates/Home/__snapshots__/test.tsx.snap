// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render the home correctly 1`] = `
.c3 {
  font-size: 2.2rem;
  font-weight: 700;
  color: #FFFFFF;
}

.c7 {
  font-size: 2.2rem;
  font-weight: 700;
  color: #000000;
}

.c2 {
  width: 100%;
  padding: 2.7rem 3.7rem;
  background: #7695EC;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: end;
  -ms-flex-pack: end;
  justify-content: end;
  width: 100%;
  margin-top: 1.6rem;
}

.c16 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  padding: 0.68rem 3rem;
  font-size: 1.6rem;
  color: #FFFFFF;
  text-transform: uppercase;
  background: #7695EC;
  border: none;
  border-radius: 0.8rem;
  -webkit-transition: 0.3s ease-in-out;
  transition: 0.3s ease-in-out;
}

.c16:disabled {
  cursor: not-allowed;
  background: #b1b1b1;
}

.c16:not(:disabled):is(:hover,:focus) {
  background: #4972e6;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.8rem;
}

.c10 {
  font-size: 1.6rem;
}

.c11 {
  width: 100%;
  padding: 0.8rem 1.12rem;
  font-size: 1.4rem;
  border: 1px solid #777777;
  border-radius: 0.8rem;
}

.c11::-webkit-input-placeholder {
  color: #CCCCCC;
}

.c11::-moz-placeholder {
  color: #CCCCCC;
}

.c11:-ms-input-placeholder {
  color: #CCCCCC;
}

.c11::placeholder {
  color: #CCCCCC;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.8rem;
}

.c13 {
  font-size: 1.6rem;
}

.c14 {
  resize: vertical;
  width: 100%;
  min-height: 7.4rem;
  padding: 0.8rem 1.12rem;
  font-size: 1.4rem;
  border: 1px solid #777777;
  border-radius: 0.8rem;
}

.c14::-webkit-input-placeholder {
  color: #CCCCCC;
}

.c14::-moz-placeholder {
  color: #CCCCCC;
}

.c14:-ms-input-placeholder {
  color: #CCCCCC;
}

.c14::placeholder {
  color: #CCCCCC;
}

.c19 {
  cursor: pointer;
  color: #FFFFFF;
}

.c5 {
  overflow: hidden;
  background: #FFFFFF;
  border: 1px solid #CCCCCC;
  border-radius: 1.6rem;
  width: 100%;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 2.4rem;
  width: 100%;
  padding: 2.4rem;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  padding: 2.7rem 3.7rem;
  background: #7695EC;
}

.c18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2.38rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 2.4rem;
}

.c20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 1.6rem;
}

.c21 {
  font-size: 1.8rem;
  font-weight: 700;
  color: #777777;
}

.c22 {
  font-size: 1.8rem;
  color: #777777;
}

.c23 {
  font-size: 1.8rem;
}

.c0 {
  max-width: 80rem;
  margin: 0 1.6rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 2.4rem;
  background: #FFFFFF;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 2.4rem;
  padding: 0 2.4rem 2.4rem;
}

@media (min-width:832px) {
  .c0 {
    margin-left: auto;
    margin-right: auto;
  }
}

<section
  class="c0 c1"
>
  <header
    class="c2"
  >
    <h1
      class="c3"
      color="secondary"
    >
      CodeLeap Network
    </h1>
  </header>
  <main
    class="c4"
  >
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        <h2
          class="c7"
          color="primary"
        >
          Whatâ€™s on your mind?
        </h2>
        <form
          class="c8"
        >
          <div
            class="c9"
          >
            <label
              class="c10"
              for=":r0:"
            >
              Title
            </label>
            <input
              class="c11"
              id=":r0:"
              placeholder="Hello world"
            />
          </div>
          <div
            class="c12"
          >
            <label
              class="c13"
              for=":r1:"
            >
              Content
            </label>
            <textarea
              class="c14"
              id=":r1:"
              placeholder="Content here"
            />
          </div>
          <div
            class="c15"
          >
            <button
              class="c16"
            >
              create
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="c5"
    >
      <div
        class="c17"
      >
        <h2
          class="c3"
          color="secondary"
        >
          My First Post at CodeLeap Network!
        </h2>
        <div
          class="c18"
        >
          <div
            class="c19"
          >
            <svg
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12l1.41 1.41L13.41 14l2.12 2.12l-1.41 1.41L12 15.41l-2.12 2.12l-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class="c19"
          >
            <svg
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m7 17.013l4.413-.015l9.632-9.54c.378-.378.586-.88.586-1.414s-.208-1.036-.586-1.414l-1.586-1.586c-.756-.756-2.075-.752-2.825-.003L7 12.583v4.43zM18.045 4.458l1.589 1.583l-1.597 1.582l-1.586-1.585l1.594-1.58zM9 13.417l6.03-5.973l1.586 1.586l-6.029 5.971L9 15.006v-1.589z"
                fill="currentColor"
              />
              <path
                d="M5 21h14c1.103 0 2-.897 2-2v-8.668l-2 2V19H8.158c-.026 0-.053.01-.079.01c-.033 0-.066-.009-.1-.01H5V5h6.847l2-2H5c-1.103 0-2 .897-2 2v14c0 1.103.897 2 2 2z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="c6"
      >
        <div
          class="c20"
        >
          <strong
            class="c21"
          >
            @
            GabrielGuedess
          </strong>
          <span
            class="c22"
          >
            an hour ago
          </span>
        </div>
        <span
          class="c23"
        >
          Text
        </span>
      </div>
    </div>
  </main>
</section>
`;
